### PR TITLE
Throw an exception when credentials have expired.

### LINKED ISSUE #409: Editing continues as an IP if the auth cookies expire
If we login with WDTK and then the cookies expire (for instance because we logout from the same account with a browser), then editing in WDTK will continue as an IP, which is not desirable.

The `assert=` parameter can be used to make sure the action fails if we are not authenticated as the given user. We should use that:
https://www.mediawiki.org/wiki/API:Assert

### PR DESCRIPTION
**Pull request by wetneb**
See https://www.mediawiki.org/wiki/API:Assert.
Fixes #409.

### CODE REVIEW

File: wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/apierrors/AssertUserFailedException.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * credentials have expired. See
+ * <a href="https://www.mediawiki.org/wiki/API:Assert">MediaWiki documentation</a>.
+ * 
+ * @author Antonin Delpeuch
+ *
+ */
+public class AssertUserFailedException extends MediaWikiApiErrorException {
+
+	public AssertUserFailedException(String errorMessage) {
+		super(MediaWikiApiErrorHandler.ERROR_NO_SUCH_ENTITY, errorMessage);
```
Comment (Tpt) [inline wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/apierrors/AssertUserFailedException.java]: Wrong error constant
```suggestion
		super(MediaWikiApiErrorHandler.ERROR_ASSERT_USER_FAILED, errorMessage);
```

File: wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/ApiConnection.java [hunk 2]
```diff
@@ -179,7 +184,7 @@
 	 */
 	@JsonIgnore
 	String password = "";
-
+
```
Comment (Tpt) [inline wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/ApiConnection.java]: training whitespace
```suggestion

```

**Comment by wetneb** [comment]
@Tpt yes I was wondering about that too! One the one hand I think it is nice that the `assert` is added by the `ApiConnection` and not the `WbEditingAction`, as this really is an authentication concern. On the other hand it does apply the parameter for all requests, even the non-editing one. That being said, in general, you do want to make sure that you are logged in for non-editing actions too: for instance, assuming you have a bot flag, this will let you request batches of 500 entities instead of 50 with `wbgetentities`.

Maybe `ApiConnection` should expose two methods (or one with a boolean parameter) to tell if we want to make the given request as a logged-in user or not. Dynamically determining that based on the query parameters inside `ApiConnection` sounds sloppy to me, there are so many different API actions that we are likely to miss some corner cases.

File: wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/apierrors/MediaWikiApiErrorHandler.java [hunk 3]
```diff
@@ -33,6 +33,7 @@
 	public final static String ERROR_INVALID_TOKEN = "badtoken";
 	public final static String ERROR_NO_SUCH_ENTITY = "no-such-entity";
 	public final static String ERROR_MAXLAG = "maxlag";
+	private static final String ERROR_ASSERT_USER_FAILED = "assertuserfailed";
```
Comment (Tpt) [inline wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/apierrors/MediaWikiApiErrorHandler.java]: It should be package private to be able to reuse it in AssertUserFailedException definition.

**Comment by Tpt** [comment]
Thank you for your comment. I believe it make sense to add the verification to all queries to keep the API simple and avoid too many options. What do you think about it?

**Comment by wetneb** [comment]
Sure, that sounds reasonable to me. If users have special needs it should always be possible to hold two `ApiConnection` objects, one logged in and the other logged out.

File: wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/apierrors/MediaWikiApiErrorHandler.java (hunk 3 above)
Comment (wetneb) [reply wdtk-wikibaseapi/src/main/java/org/wikidata/wdtk/wikibaseapi/apierrors/MediaWikiApiErrorHandler.java]: Oops, sorry! I have made it public to match the rest of the codes above.

**Comment by Tpt** [comment]
Thank you! Do you want to get a release quickly for inclusion in OpenRefine?

**Comment by wetneb** [comment]
That would be helpful, thanks. I thought maybe I could squeeze in #268 too?

**Comment by Tpt** [comment]
> I thought maybe I could squeeze in #268 too?

It would be great!

**Pull request merged**